src/wallet/test/db_tests.cpp

```diff
@@ -6,13 +6,62 @@
 
 #include <test/util/setup_common.h>
 #include <util/fs.h>
+#include <util/translation.h>
+#ifdef USE_BDB
 #include <wallet/bdb.h>
+#endif
+#ifdef USE_SQLITE
+#include <wallet/sqlite.h>
+#endif
+#include <wallet/test/util.h>
+#include <wallet/walletutil.h> // for WALLET_FLAG_DESCRIPTORS
 
 #include <fstream>
 #include <memory>
 #include <string>
 
+inline std::ostream& operator<<(std::ostream& os, const std::pair<const SerializeData, SerializeData>& kv)
+{
+    Span key{kv.first}, value{kv.second};
+    os << "(\"" << std::string_view{reinterpret_cast<const char*>(key.data()), key.size()} << "\", \""
+       << std::string_view{reinterpret_cast<const char*>(key.data()), key.size()} << "\")";
+    return os;
+}
+
 namespace wallet {
+
+static bool WriteBytes(DatabaseBatch& batch, Span<const std::byte> key, Span<const std::byte> value,
+                       bool overwrite = true)
+{
+    return batch.Write(DataStream{key}, DataStream{value}, overwrite);
+}
```

Comment:
In commit "wallet: Add GetPrefixCursor to DatabaseBatch" (13476fe7bebdbf51e09821850b2c808c8ecf116a)

Maybe add a comment above the batch.Write() call like "Convert the key and value to DataStream objects in order to bypass serialization. We want raw bytes to be written to the database, not serialized byte strings. The DatabaseBatch::Write template method normally serializes its arguments, but because DataStream has a Serialize method that does concatenation instead of serialization, it can be used to bypass serialization."

I was very surprised that this worked and it took me a while to figure out what was happening. I do think it would probably be more straightforward to call the private `DatabaseBatch::WriteKey` method instead of `DatabaseBatch::Write` which is why I added the friend class in my earlier suggestion. But I guess this approach should be ok as long it is explained.

It does seem pretty crazy to me that DataStream has a Serialize method that concatenates instead of serializing, and can't be unserialized. But I guess it is handy in this situation and maybe others.

Reply:
> Is the same not true for span itself? 

No because `Span<char>` is fixed length and can be deserialized, while `DataStream` is variable length and can't be deserialized.

Reasons why I think it would be good to get rid of `DataStream::Serialize` method:

- `DataStream` has no `Unserialize` method and no reasonable way of adding one that would be consistent with the existing `Serialize` implementation.
- The `Serialize` method has surprising behavior. If `DataStream` was going to have a serialize method I would expect to behave same was as serialize methods as other variable length objects (strings, vectors, maps) and be prefixed with a size field and be deserializable into an empty object. I suggested adding a comment here because code using it was not working the way I expected it to, and it took me a while to figure out how it was working at all.
- No need or use-case for this `Serialize` method. If we want a concatenation operator, `+=` should work perfectly well. `<<` is good to use for serialization and formatting, and should in principle be reversible with `>>`. Concatenation is a different thing.

If I remove `DataStream::Serialize` method in master there are only a few compilation errors, so I think I should be able to make a simple PR to clean things up

Reply:
I think @MarcoFalke's point was that `Span`s are serialized without length prefixes, so this function could just pass `key` and `value` directly to `DatabaseBatch::Write` rather than going through a `DataStream`, as long as there is a `Serialize` method for `Span<const std::byte>`. This seems like a more intuitive solution, so I've implemented that.

Reply:
Oops, I guess I was objecting to a point Marco wasn't making. But I think I would still like to avoid overloading Serialize for std::byte spans. The whole point of std::byte is that it's supposed to be a very safe type which requires you to be deliberate and explicit about conversions. So I'm not sure it would be great to serialize std::byte spans (or types which can be converted to byte spans) as raw bytes without requiring a more explicit cast.

Maybe it would make sense to change serialize.h in the future, but I think it would be best to leave it alone in this PR. Would suggest reverting serialize.h and just doing:

```diff
--- a/src/wallet/test/db_tests.cpp
+++ b/src/wallet/test/db_tests.cpp
@@ -30,12 +30,6 @@ inline std::ostream& operator<<(std::ostream& os, const std::pair<const Serializ
 
 namespace wallet {
 
-static bool WriteBytes(DatabaseBatch& batch, Span<const std::byte> key, Span<const std::byte> value,
-                       bool overwrite = true)
-{
-    return batch.Write(key, value, overwrite);
-}
-
 static Span<const std::byte> StringBytes(std::string_view str)
 {
     return AsBytes<const char>({str.data(), str.size()});
@@ -203,7 +197,7 @@ BOOST_AUTO_TEST_CASE(db_cursor_prefix_byte_test)
     for (const auto& database : TestDatabases(m_path_root)) {
         std::unique_ptr<DatabaseBatch> batch = database->MakeBatch();
         for (const auto& [k, v] : {e, p, ps, f, fs, ff, ffs}) {
-            WriteBytes(*batch, k, v);
+            batch->Write(MakeUCharSpan(k), MakeUCharSpan(v));
         }
         CheckPrefix(*batch, StringBytes(""), {e, p, ps, f, fs, ff, ffs});
         CheckPrefix(*batch, StringBytes("prefix"), {p, ps});

```

Reply:
Good point, done as suggested.

Reply:
I think I agree that current `Serialize(Span<char>)` overloads could be a little dangerous:

https://github.com/bitcoin/bitcoin/blob/6a560aceb75e618f3106a8850e053cd8de87616a/src/serialize.h#L202-L205

Because if a variable-length container could be implicitly converted to span of these types, it might incorrectly match one of these overloads and be serialized as a fixed length span without a length prefix, and no longer be deserializable as the original container.

If that is a real problem, we could probably fix it by replacing those overloads with:

```c++
template<typename Stream, typename T> inline void Serialize(Stream& s, const Span<T>& span);
template<typename Stream, typename T> inline void Unserialize(Stream& s, Span<T>& span);
```

Since these would only match actual `Span` objects, and would not match arguments implicitly convertible to Span. I think this would be as safe as your `MakeSerSpan` idea and not require introducing a new type.

I also agree that the `Span<char>` char type is different `char[N]` type because one is a fixed length type and the other just has fixed length objects (`subspan` method returns a new object, it doesn't change the size of an existing object). But the point about deserialization applies to both of cases, because you can deserialize both `char[N]` objects and `Span<char>` objects without a length prefix.